### ISSUE TITLE
[AMDGPU][MC][GFX11] v_fmac_f16_e64 with non-default op_sel is disassembled incorrectly

### ISSUE THREAD

**Issue by dpreobra** [llvm/llvm-project#59468]
op_sel is disassembled as if v_fmac_f16_e64 has 3 src operands.
An example of failed test:

    0x05,0x08,0x36,0xd5,0x01,0x05,0x02,0x00

Expected result:

    v_fmac_f16_e64 v5, v1, v2 op_sel:[1,0,0]

Actual output:

    v_fmac_f16_e64 v5, v1, v2 op_sel:[1,0,0,0]

**Comment by dpreobra** [comment]
I agree that the code with src2.opsel=1 may be legal, and we probably can make it accepted by llvm. But I do not see how it relates to op_sel:[...] printing. The instruction has no src2, so the bit should be ignored.

I think we should be consistent: all other instructions have op_sel printed with the number of elements that match the number of operands, but v_fmac_f16_e64 is an exception.

**Comment by Sisyph** [comment]
You're right, src2.opsel =1 is a separate issue. Printing consistently is good. I think the suffix syntax .l/.h is superior for clarity, but we are not quite ready to remove op_sel. Which instructions with 2 source operands and op_sel are you considering; I'm having a hard time finding one like that? The reason I prefer always printing 4 bits of opsel is a) the instruction encoding always has 4 bits and b) I don't like that the 3rd bit could refer to either src2 or dst if we had 3 bits in the op_sel field. This is a consequence of the difference between operand order and the operand order within op_sel:[ ]

**Comment by dpreobra** [comment]
> Which instructions with 2 source operands and op_sel are you considering

There are a lot of GFX11 instructions with 2 source operands and op_sel, but most of them do not support op_sel yet. These are mostly instructions promoted from VOP1, VOP2 and VOPC. Right now we have only 5 instructions which support op_sel:

    v_add_nc_i16
    v_add_nc_u16
    v_pack_b32_f16
    v_sub_nc_i16
    v_sub_nc_u16

An example taken from our lit tests:

    # GFX10: v_add_nc_u16 v5, v1, v2 op_sel:[1,1,1]  ; encoding: [0x05,0x58,0x03,0xd7,0x01,0x05,0x02,0x00]
    0x05,0x58,0x03,0xd7,0x01,0x05,0x02,0x00

I agree with your analysis; the current op_sel syntax is counterintuitive, however it has been used for a long time for GFX10. 
I do not mind always printing 4 elements, but should not we enforce using the same syntax for assembler as well?

If you are going to switch to .h/.l syntax in the near future, then I think all these problems look insignificant and this issue may be closed.

**Comment by Sisyph** [comment]
Thanks for the example. I am definitely working on areas related to op_sel and suffix support. It is still somewhat in flux how soon support for suffixes on VGPRs, SGPRs or immediates can be ready, though VGPRs will likely be first, and I would estimate a few months. I am fine with it however you want to resolve this issue for FMAC (3 bits, 4, or do nothing).

**Comment by jwanggit86** [comment]
Quoted from above:
> If you are going to switch to .h/.l syntax in the near future, then I think all these problems look insignificant, and this issue may be closed.

It looks like that syntax has indeed switched to .h/.l suffixes.

The instruction `v_fmac_f16_e64` no longer allows op_sel:
```
$ cat bad.s
v_fmac_f16_e64 v5, v1, v2 op_sel:[1,0,0]
v_fmac_f16_e64 v5, v1, v2

$ ../../build/bin/llvm-mc -triple=amdgcn -mcpu=gfx1100 -show-encoding bad.s
bad.s:1:27: error: not a valid operand.
v_fmac_f16_e64 v5, v1, v2 op_sel:[1,0,0]
                          ^
        v_fmac_f16_e64 v5, v1, v2               ; encoding: [0x05,0x00,0x36,0xd5,0x01,0x05,0x02,0x00]
```
The first instruction has op_sel and is rejected by the assembler. The 2nd does not have op_sel and is assembled successfully.

If we use the encoding of the 2nd instruction, but change the 2nd byte from 0x00 to 0x08 (the 4 OPSEL bits would be 0001)
and run it with the disassembler, we get:
```
$ cat bad.txt
0x05,0x08,0x36,0xd5,0x01,0x05,0x02,0x00

$ ../../build/bin/llvm-mc -triple=amdgcn -mcpu=gfx1100 -show-encoding -disassemble bad.txt
        v_fmac_f16_e64 v5, v1, v2               ; encoding: [0x05,0x00,0x36,0xd5,0x01,0x05,0x02,0x00]
```
It can be disassembled, but the resulted instruction has no op_sel and the encoding shows that the 2nd byte is 0x00.

So it appears the original problem no longer exists with this instruction. Please let me know your thoughts before I close this issue.

**Comment by jwanggit86** [comment]
Closing this issue as the original problem no longer exists.
